neuralmonkey/encoders/conv_s2s_encoder.py

```diff
@@ -0,0 +1,189 @@
+                convolutions = self.residual_conv(
+                    convolutions, "encoder_conv_{}".format(layer))
+
+            self.states = convolutions + linear(self.ordered_embedded_inputs,
+                                                self.conv_features)
+            # This state concatenation is not based on any paper, but was
+            # tested empirically
+            self.encoded = tf.reduce_max(self.states, axis=1)
+    # pylint: enable=too-many-arguments
+
+    @tensor
+    def _attention_tensor(self) -> tf.Tensor:
+        return dropout(self.states, self.dropout_keep_prob, self.train_mode)
+
+    @tensor
+    def _attention_mask(self) -> tf.Tensor:
+        # TODO tohle je proti OOP prirode
+        return self.input_sequence.mask
+
+    @tensor
+    def states_mask(self) -> tf.Tensor:
+        return self.input_sequence.mask
+
+    @tensor
+    def order_embeddings(self) -> tf.Tensor:
+        # initialization in the same way as in original CS2S implementation
+        with tf.variable_scope("input_projection"):
+            return tf.get_variable(
+                "order_embeddings", [self.input_sequence.max_length,
+                                     self.input_sequence.embedding_sizes[0]],
+                initializer=tf.random_normal_initializer(stddev=0.1))
+
+    @tensor
+    def ordered_embedded_inputs(self) -> tf.Tensor:
+        # shape (batch, time, embedding size)
+        ordering_additive = tf.expand_dims(self.order_embeddings, 0)
+        batch_max_len = tf.shape(self.input_sequence.data)[1]
+        clipped_ordering_embed = ordering_additive[:, :batch_max_len, :]
+
+        return self.input_sequence.data + clipped_ordering_embed
+
+    def residual_conv(self, input_signals, name):
+        with tf.variable_scope(name):
+            # initialized as described in the paper
+            init_deviat = np.sqrt(4 / self.conv_features)
+            convolution_filters = tf.get_variable(
+                "convolution_filters",
+                [self.kernel_width, self.conv_features,
+                 2 * self.conv_features],
+                initializer=tf.random_normal_initializer(stddev=init_deviat))
+
+            bias = tf.get_variable(
+                name="conv_bias",
+                shape=[2 * self.conv_features],
+                initializer=tf.zeros_initializer())
+
+            conv = (tf.nn.conv1d(input_signals, convolution_filters, 1, "SAME")
+                    + bias)
+
+            return glu(conv) + input_signals
+
+    # pylint: disable=no-self-use
+    @tensor
+    def train_mode(self):
+        # scalar tensor
+        return tf.placeholder(tf.bool, shape=[], name="mode_placeholder")
+    # pylint: enable=no-self-use
+
+    def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
+        fd = self.input_sequence.feed_dict(dataset, train)
+        fd[self.train_mode] = train
+
+        return fd
+
+
+class SentenceEncoder(ConvolutionalSentenceEncoder):
```

Comment:
Ale ještě bych nějak rozumněji pojmenoval ty třídy. Protože to `Sentence` by IMHO mělo být v názvu tohodle syntactic sugaru. Chtělo by to nějaký název, ze kterého někoku, kdo se pohybuje v oboru, bude zřejmé, o co se jedná. Co to vlastně je?  Fully convolutional seq2seq encoder? Facebook's convolutional encoder?